server/mqtt.go

```diff
@@ -2563,8 +2628,14 @@ func (as *mqttAccountSessionManager) loadRetainedMessagesForSubject(rms map[stri
 			log.Warnf("failed to decode retained message for subject %q: %v", loadSubject, err)
 			continue
 		}
-		rms[subject] = &rm
+
+		// Add the loaded retained message to the cache.
+		ref.loaded = &rm
+		ref.loaded.expiresFromCache = time.Now().Add(mqttRetainedCacheTTL)
```

Comment:
Again, this should be protected, but also assumes that this message will not expire by the time we need it.